### PR TITLE
GH-36776: [C++] Make ListArray::FromArrays() handle sliced offsets Arrays containing nulls

### LINKED ISSUE #36776: [C++] array_nested.cc: CleanListOffsets() assumes offsets.offset() is always 0
### Describe the bug, including details regarding any error messages, version, and platform.

`CleanListOffsets` (code below) copies the validity buffer slice starting from 0, but that can lead to copying the wrong bits when `offsets.offset()` is greater than 0.

```cpp
    ARROW_ASSIGN_OR_RAISE(
        auto clean_validity_buffer,
        offsets.null_bitmap()->CopySlice(0, bit_util::BytesForBits(num_offsets - 1)));
```

https://github.com/apache/arrow/blob/3f0b62038d7ddac090db316722f5a1695e85701b/cpp/src/arrow/array/array_nested.cc#L55-L99

The fix is using`CopyBitmap` which knows how to handle offsets at the bit granularity plus some extra adaptations.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

It fixes the issue described in #36776 

### What changes are included in this PR?

The fix and a simplifications of the interaction between `ListArrayFromArrays` and `CleanOffsets` which is rather involved.

### Are these changes tested?

Yes. I added a test that reproduces the issue before adding the fixes.
* Closes: #36776

### CODE REVIEW

**Comment by felipecrv** [comment]
@pitrou review this commit by commit as the whole change might be confusing.

File: cpp/src/arrow/array/array_list_test.cc [hunk 1]
```diff
@@ -233,6 +233,26 @@ class TestListArray : public ::testing::Test {
                                                  expected->null_bitmap()));
   }
 
+  void TestFromArraysWithSlicedNullBitmap() {
```
Comment (pitrou) [inline cpp/src/arrow/array/array_list_test.cc:236]: For the record, do we already have a test with sliced offsets but no validity bitmap?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a lot for noticing and fixing this @felipecrv .

File: cpp/src/arrow/array/array_list_test.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/array/array_list_test.cc:236]: I don't think so. I will add one.

File: cpp/src/arrow/array/array_list_test.cc [hunk 2]
```diff
@@ -233,6 +233,26 @@ class TestListArray : public ::testing::Test {
...
+    std::vector<offset_type> offsets = {-1, -1, 0, 1, 1, 3};
```
[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/array/array_list_test.cc:237]: Added to the two versions of tests I added.

**Review by pitrou — APPROVED** [review]

**Pull request merged**